weave/trace_server/project_version/project_version.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import threading
+from collections.abc import Callable
+from typing import Any, Optional
+
+import ddtrace
+from cachetools import LRUCache
+
+from weave.trace_server.project_version.clickhouse_project_version import (
+    get_project_data_residence,
+)
+from weave.trace_server.project_version.types import (
+    CallsStorageServerMode,
+    ProjectDataResidence,
+    ReadTable,
+    WriteTarget,
+)
+
+logger = logging.getLogger(__name__)
+
+PER_REPLICA_CACHE_SIZE = 10_000
+
+_resolver: Optional["TableRoutingResolver"] = None
+_resolver_lock = threading.Lock()
+
+
+def init_resolver(
+    ch_client_factory: Callable[[], Any],
+    cache_size: int = PER_REPLICA_CACHE_SIZE,
+) -> None:
+    """Initialize the project version resolver.
+
+    Call this once at application startup. Subsequent calls are ignored
+    (the first initialization wins).
+
+    Args:
+        ch_client_factory: Callable that returns a ClickHouse client.
+        cache_size: Size of the in-memory cache (defaults to 10,000).
+    """
+    global _resolver
+    with _resolver_lock:
+        if _resolver is None:
+            _resolver = TableRoutingResolver(
+                ch_client_factory=ch_client_factory,
+                cache_size=cache_size,
+            )
+
+
+def resolve_read_table(project_id: str) -> ReadTable:
+    if _resolver is None:
+        raise RuntimeError("init_resolver() must be called before resolve_read_table()")
+    return _resolver.resolve_read_table(project_id)
+
+
+def resolve_write_target(project_id: str) -> WriteTarget:
+    if _resolver is None:
+        raise RuntimeError(
+            "init_resolver() must be called before resolve_write_target()"
+        )
+    return _resolver.resolve_write_target(project_id)
+
+
+class TableRoutingResolver:
```

Comment:
Great! this Router contains the logic i was hoping for. a clear delineation of concerns